File: kdb-recovery/src/commonTest/kotlin/dev/kdb/recovery/RestoreTest.kt
package dev.kdb.recovery

import dev.kdb.codec.KdbHash
import dev.kdb.codec.KdbTimestamp
import dev.kdb.codec.KdbUuid
import dev.kdb.compression.Crc32
import dev.kdb.document.DocumentTree
import dev.kdb.document.KdbCommit
import dev.kdb.document.KdbOp
import dev.kdb.integrity.scanVerifiedCommits
import dev.kdb.storage.CompressionCodec
import dev.kdb.storage.PlatformIoShim
import dev.kdb.storage.io.SegmentNameBuilder
import dev.kdb.storage.mem.InMemoryPlatformIoShim
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest

class RestoreTest {
    @Test
    fun restoreFromBackupOnlyRebuildsCleanly() =
        runTest {
            val ns = "ns1"
            val backup = newDirShim()
            val c0 = buildCommit(ns, null)
            val c1 = buildCommit(ns, c0.hash)
            appendSegment(backup, ns, 0, rawFrame(c0), rawFrame(c1))

            val out = newDirShim()
            val result = hybridRestore(listOf(Source("backup", backup)), ns, CompressionCodec.NONE, out)
            assertEquals(2, result.appliedCount)
            assertTrue(result.missingHashes.isEmpty())

            val restored = scanVerifiedCommits(out, ns, CompressionCodec.NONE)
            assertEquals(2, restored.size)
            assertTrue(restored.containsKey(c0.hash.toHex()))
            assertTrue(restored.containsKey(c1.hash.toHex()))
        }

    @Test
    fun hybridRestoreLocalAheadOfBackup() =
        runTest {
            val ns = "ns1"
            val c0 = buildCommit(ns, null)
            val c1 = buildCommit(ns, c0.hash)
            val c2 = buildCommit(ns, c1.hash)
            val c3 = buildCommit(ns, c2.hash) // will be torn in the local copy

            val backup = newDirShim()
            appendSegment(backup, ns, 0, rawFrame(c0), rawFrame(c1)) // stale: missing c2, c3

            val local = newDirShim()
            val torn = rawFrame(c3).copyOfRange(0, 8)
            appendSegment(local, ns, 0, rawFrame(c0), rawFrame(c1), rawFrame(c2), torn)

            val out = newDirShim()
            val result = hybridRestore(listOf(Source("local", local), Source("backup", backup)), ns, CompressionCodec.NONE, out)
            assertEquals(3, result.appliedCount, "expected 3 applied commits (c0,c1,c2), got $result")
            assertTrue(result.missingHashes.isEmpty())

            val restored = scanVerifiedCommits(out, ns, CompressionCodec.NONE)
            for (h in listOf(c0.hash, c1.hash, c2.hash)) {
                assertTrue(restored.containsKey(h.toHex()), "expected restored log to contain ${h.toHex()}")
            }
            assertFalse(restored.containsKey(c3.hash.toHex()), "restored log must not contain the torn, unverified commit c3")
        }

    @Test
    fun hybridRestoreFillsGapFromPeer() =
        runTest {
            val ns = "ns1"
            val c0 = buildCommit(ns, null)
            val c1 = buildCommit(ns, c0.hash) // missing from both local and backup
            val c2 = buildCommit(ns, c1.hash)

            val local = newDirShim()
            appendSegment(local, ns, 0, rawFrame(c0))
            appendSegment(local, ns, 1, rawFrame(c2)) // c2's parent c1 is absent

            val withoutPeer = hybridRestore(listOf(Source("local", local)), ns, CompressionCodec.NONE, newDirShim())
            assertTrue(withoutPeer.missingHashes.isNotEmpty(), "expected missing hashes without the peer source, got $withoutPeer")

            val peer = newDirShim()
            appendSegment(peer, ns, 0, rawFrame(c1))

            val out = newDirShim()
            val result = hybridRestore(listOf(Source("local", local), Source("peer", peer)), ns, CompressionCodec.NONE, out)
            assertEquals(3, result.appliedCount)
            assertTrue(result.missingHashes.isEmpty(), "expected the peer source to fill the gap, got $result")
        }

    @Test
    fun restoreNeverTrustsUnverifiedLocalFrames() =
        runTest {
            val ns = "ns1"
            val c0 = buildCommit(ns, null)
            val c1 = buildCommit(ns, c0.hash)

            val local = newDirShim()
            val corrupt = rawFrame(c1).copyOf()
            corrupt[19] = (corrupt[19].toInt() xor 0xFF).toByte() // CRC mismatch, not truncation
            appendSegment(local, ns, 0, rawFrame(c0), corrupt)

            val out = newDirShim()
            val result = hybridRestore(listOf(Source("local", local)), ns, CompressionCodec.NONE, out)
            assertEquals(1, result.appliedCount, "expected only the verified commit c0 to be restored, got $result")

            val restored = scanVerifiedCommits(out, ns, CompressionCodec.NONE)
            assertFalse(restored.containsKey(c1.hash.toHex()), "restore must never apply a commit whose frame failed CRC verification")
        }
}

private fun newDirShim(): PlatformIoShim = InMemoryPlatformIoShim()

private fun buildCommit(ns: String, parent: KdbHash?): KdbCommit =
    KdbCommit.build(
        parentHashes = if (parent != null) listOf(parent) else emptyList(),
        namespaceId = ns,
        transactionId = KdbUuid.random(),
        timestamp = KdbTimestamp.now(),
        authorNodeId = KdbUuid.random(),
        operations = listOf(KdbOp.Write(KdbUuid.random(), "{}")),
        documentTreeHash = DocumentTree.EMPTY.treeHash,
        schemaHash = null,
        message = "test",
    )

private fun rawFrame(c: KdbCommit): ByteArray {
    val payload = c.toPayloadBytes()
    val out = ByteArray(16 + payload.size)
    out[0] = 0x4B; out[1] = 0x44; out[2] = 0x42; out[3] = 0x50
    writeIntBe(out, 4, payload.size)
    writeIntBe(out, 8, payload.size)
    writeIntBe(out, 12, Crc32.of(payload))
    payload.copyInto(out, 16)
    return out
}

private fun writeIntBe(a: ByteArray, o: Int, v: Int) {
    a[o] = (v ushr 24).toByte(); a[o + 1] = (v ushr 16).toByte()
    a[o + 2] = (v ushr 8).toByte(); a[o + 3] = v.toByte()
}

private suspend fun appendSegment(shim: PlatformIoShim, ns: String, seq: Long, vararg frames: ByteArray) {
    val name = SegmentNameBuilder.deltaSequenced(ns, seq)
    for (f in frames) shim.appendToSegment(name, f)
}

File: go/kdb/recovery/restore_test.go
package recovery

import (
	"testing"

	"github.com/limidus/kdb/go/kdb/codec"
	"github.com/limidus/kdb/go/kdb/document"
	"github.com/limidus/kdb/go/kdb/integrity"
	"github.com/limidus/kdb/go/kdb/storage"
	"github.com/limidus/kdb/go/kdb/storage/delta"
	storio "github.com/limidus/kdb/go/kdb/storage/io"
)

func newDirShim(t *testing.T) storage.PlatformIOShim {
	t.Helper()
	root := t.TempDir()
	store, err := storio.NewOSByteStore(storio.PlatformIOConfig{RootDirectory: &root})
	if err != nil {
		t.Fatal(err)
	}
	return storio.NewFileBackedPlatformIO(storio.PlatformIOConfig{RootDirectory: &root, FsyncOnFlush: true}, store)
}

func buildCommit(t *testing.T, ns string, parent *codec.Hash) document.Commit {
	t.Helper()
	tx, _ := codec.RandomUUID()
	author, _ := codec.RandomUUID()
	docID, _ := codec.RandomUUID()
	var parents []codec.Hash
	if parent != nil {
		parents = []codec.Hash{*parent}
	}
	c := document.Commit{
		ParentHashes:     parents,
		NamespaceID:      ns,
		TransactionID:    tx,
		Timestamp:        codec.TimestampNow(),
		AuthorNodeID:     author,
		Operations:       []document.Op{document.WriteOp{DocID: docID, Patch: "{}"}},
		DocumentTreeHash: document.EmptyDocumentTree().TreeHash,
		Message:          "test",
	}
	h, err := document.ComputeCommitHash(c)
	if err != nil {
		t.Fatal(err)
	}
	c.Hash = h
	return c
}

func rawFrame(t *testing.T, c document.Commit) []byte {
	t.Helper()
	payload, err := c.ToPayloadBytes()
	if err != nil {
		t.Fatal(err)
	}
	frame, err := (delta.PageCodec{}).Frame(payload, storage.CompressionNone)
	if err != nil {
		t.Fatal(err)
	}
	return frame
}

func appendSegment(t *testing.T, shim storage.PlatformIOShim, ns string, seq int64, frames ...[]byte) {
	t.Helper()
	name := storio.SegmentNameBuilder.DeltaSequenced(ns, seq)
	for _, f := range frames {
		if _, err := shim.AppendToSegment(name, f); err != nil {
			t.Fatal(err)
		}
	}
}

func TestRestoreFromBackupOnlyRebuildsCleanly(t *testing.T) {
	const ns = "ns1"
	backup := newDirShim(t)
	c0 := buildCommit(t, ns, nil)
	c1 := buildCommit(t, ns, &c0.Hash)
	appendSegment(t, backup, ns, 0, rawFrame(t, c0), rawFrame(t, c1))

	out := newDirShim(t)
	result, err := HybridRestore([]Source{{Label: "backup", Shim: backup}}, ns, storage.CompressionNone, out)
	if err != nil {
		t.Fatal(err)
	}
	if result.AppliedCount != 2 || len(result.MissingHashes) != 0 {
		t.Fatalf("unexpected result: %+v", result)
	}

	restored, err := integrity.ScanVerifiedCommits(out, ns)
	if err != nil {
		t.Fatal(err)
	}
	if len(restored) != 2 || !hasHash(restored, c0.Hash) || !hasHash(restored, c1.Hash) {
		t.Fatalf("restored log missing expected commits: %+v", restored)
	}
}

func TestHybridRestoreLocalAheadOfBackup(t *testing.T) {
	const ns = "ns1"
	c0 := buildCommit(t, ns, nil)
	c1 := buildCommit(t, ns, &c0.Hash)
	c2 := buildCommit(t, ns, &c1.Hash)
	c3 := buildCommit(t, ns, &c2.Hash) // will be torn in the local copy

	backup := newDirShim(t)
	appendSegment(t, backup, ns, 0, rawFrame(t, c0), rawFrame(t, c1)) // stale: missing c2, c3

	local := newDirShim(t)
	torn := rawFrame(t, c3)[:8]
	appendSegment(t, local, ns, 0, rawFrame(t, c0), rawFrame(t, c1), rawFrame(t, c2), torn)

	out := newDirShim(t)
	result, err := HybridRestore(
		[]Source{{Label: "local", Shim: local}, {Label: "backup", Shim: backup}},
		ns, storage.CompressionNone, out,
	)
	if err != nil {
		t.Fatal(err)
	}
	if result.AppliedCount != 3 {
		t.Fatalf("expected 3 applied commits (c0,c1,c2), got %+v", result)
	}
	if len(result.MissingHashes) != 0 {
		t.Fatalf("expected no missing hashes, got %v", result.MissingHashes)
	}

	restored, err := integrity.ScanVerifiedCommits(out, ns)
	if err != nil {
		t.Fatal(err)
	}
	for _, h := range []codec.Hash{c0.Hash, c1.Hash, c2.Hash} {
		if !hasHash(restored, h) {
			t.Fatalf("expected restored log to contain %s", h.Hex())
		}
	}
	if hasHash(restored, c3.Hash) {
		t.Fatal("restored log must not contain the torn, unverified commit c3")
	}
}

func TestHybridRestoreFillsGapFromPeer(t *testing.T) {
	const ns = "ns1"
	c0 := buildCommit(t, ns, nil)
	c1 := buildCommit(t, ns, &c0.Hash) // missing from both local and backup
	c2 := buildCommit(t, ns, &c1.Hash)

	local := newDirShim(t)
	appendSegment(t, local, ns, 0, rawFrame(t, c0))
	appendSegment(t, local, ns, 1, rawFrame(t, c2)) // c2's parent c1 is absent

	withoutPeer, err := HybridRestore([]Source{{Label: "local", Shim: local}}, ns, storage.CompressionNone, newDirShim(t))
	if err != nil {
		t.Fatal(err)
	}
	if len(withoutPeer.MissingHashes) == 0 {
		t.Fatalf("expected missing hashes without the peer source, got %+v", withoutPeer)
	}

	peer := newDirShim(t)
	appendSegment(t, peer, ns, 0, rawFrame(t, c1))

	out := newDirShim(t)
	result, err := HybridRestore(
		[]Source{{Label: "local", Shim: local}, {Label: "peer", Shim: peer}},
		ns, storage.CompressionNone, out,
	)
	if err != nil {
		t.Fatal(err)
	}
	if result.AppliedCount != 3 || len(result.MissingHashes) != 0 {
		t.Fatalf("expected the peer source to fill the gap, got %+v", result)
	}
}

func TestRestoreNeverTrustsUnverifiedLocalFrames(t *testing.T) {
	const ns = "ns1"
	c0 := buildCommit(t, ns, nil)
	c1 := buildCommit(t, ns, &c0.Hash)

	local := newDirShim(t)
	corrupt := append([]byte(nil), rawFrame(t, c1)...)
	corrupt[19] ^= 0xFF // CRC mismatch, not truncation - a frame that "fits" but is wrong
	appendSegment(t, local, ns, 0, rawFrame(t, c0), corrupt)

	out := newDirShim(t)
	result, err := HybridRestore([]Source{{Label: "local", Shim: local}}, ns, storage.CompressionNone, out)
	if err != nil {
		t.Fatal(err)
	}
	if result.AppliedCount != 1 {
		t.Fatalf("expected only the verified commit c0 to be restored, got %+v", result)
	}
	restored, err := integrity.ScanVerifiedCommits(out, ns)
	if err != nil {
		t.Fatal(err)
	}
	if hasHash(restored, c1.Hash) {
		t.Fatal("restore must never apply a commit whose frame failed CRC verification")
	}
}

func hasHash(m map[string]document.Commit, h codec.Hash) bool {
	_, ok := m[h.Hex()]
	return ok
}
